### ISSUE TITLE
Size limit issue

### ISSUE THREAD

**Issue by metallikat36** [dsifford/academic-bloggers-toolkit#378]
Hi Derek. I know I've bugged you previously about my posts giving me "Request Entity Too Large" when I try to update them. My server has a Post_max_size of 128MB. Can you give me some kind of idea how much of that space could be used by having citations? I definitely notice that citations are a major factor in when my post finally gets so big as to break (and a post of exclusively citations with no text breaks for me around 250 citations, if I recall my past experiments correctly), but it's kind of hard to believe that citations would take up so much data. Also, are there any possible improvements with the gutenberg whatever changes that are coming that I might expect my posts to be able to be longer and still go through?

**Comment by dsifford** [comment]
> Can you give me some kind of idea how much of that space could be used by having citations?

This depends on a few factors. Probably the biggest factor is the total number of times each citation is used in the document. 

**To give a quick rundown:**

(note: let `n` = total number of individual references)

Citation data is stored as JSON as post meta. This (among a couple of other things) has two top-level fields that contain citation metadata: One stores all citations individually by its unique identifier (so size here is `n`). The other stores information about each citation in the document indexed by occurrence in the document itself. This is where the size can quickly become larger. 

Each time a reference is used in the document, the entire citation object must be fully copied into the citation index. If you have one reference and you cite it 50 times, you will have 51 copies of that metadata in the data JSON.

You might be asking yourself, "isn't that non-optimal? Couldn't that be improved by instead passing the identifier to the citation object, rather that the full reference metadata?"... The answer is yes. But that's not in my control.

> it's kind of hard to believe that citations would take up so much data

Believe it! There's quite a lot of data to keep track of in order to properly manage document citations.

> Also, are there any possible improvements with the gutenberg whatever changes that are coming that I might expect my posts to be able to be longer and still go through?

The future of this plugin is still not certain. As of the time of this writing, there is still no way for it to work with Gutenberg. I'm hoping for that to change, but it has not yet.

Your historic posts should not be in jeopardy though. Those should transfer fine.



Hope that helps.

**Comment by metallikat36** [comment]
FYI, I read that there will be a plugin that disables gutenberg.

**Comment by metallikat36** [comment]
Would something like the following be possible? Instead of having all the citation data stored in a post, maybe instead just have a call to some function stored which would hopefully take up less data? Then when you actually view the post in real time, those functions are called and then they pull the info from pubmed or wherever? In other words, some kind of lazy-loading capability that won't pull data until a user clicks on a citation to see a tooltip, or tries to expand the reference list. Of course, that would not work for references which are manually inputted, but that is probably relatively few overall.

And I wonder if there is anyway to divide the Update Post or Preview Posts actions into multiple requests of smaller size?

Maybe I am just talking rubbish, as I am not a programmer. But just throwing out some concepts.

**Comment by dsifford** [comment]
Good thoughts, but unfortunately those wouldn't solve the issue.

Data stored _in_ the actual post is not much more than what you see when you're typing it. Tooltip data, for example, is just a reference to the corresponding item in the bibliography. When a post is loaded for a user, the post is quickly iterated and the tooltip content for each citation is generated during runtime by copying the html of the reference list item to a hidden data attribute. In other words, if there are three items contained in a single citation, the only data that is stored in the actual post at that position is `<span class="abt-citation" data-reflist="["12345","67890","24680"]>1</span>`. (assume those numbers are IDs of references contained in the bibliography).

Long story short: The issue is _exclusively_ the post metadata. The size of the visible post is not the problem whatsoever. Unfortunately, for reasons already explained above, I'm currently forced to keep that metadata in the exact shape that it is in for the processor to work properly.
